lib/Dialect/FIRRTL/Transforms/LowerGroups.cpp

```diff
@@ -0,0 +1,391 @@
+//===- LowerGroups.cpp - Lower Optiona Groups to Instances ------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// This pass lowers FIRRTL optional groups to modules and instances.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLUtils.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Mutex.h"
+#include "llvm/Support/RWMutex.h"
+
+#define DEBUG_TYPE "firrtl-lower-groups"
+
+using namespace circt;
+using namespace firrtl;
+
+class LowerGroupsPass : public LowerGroupsBase<LowerGroupsPass> {
+  /// Safely build a new module with a given namehint.  This handles geting a
+  /// lock to modify the top-level circuit.
+  FModuleOp buildNewModule(OpBuilder &builder, Location location,
+                           Twine namehint, SmallVectorImpl<PortInfo> &ports);
+
+  /// Function to process each module.
+  void runOnModule(FModuleOp moduleOp);
+
+  /// Entry point for the function.
+  void runOnOperation() override;
+
+  /// Indicates exclusive access to modify the circuitNamespace and the circuit.
+  llvm::sys::SmartMutex<true> *circuitMutex;
+
+  /// A circuit-level namespace.
+  CircuitNamespace *circuitNamespace;
+};
+
+/// Multi-process safe function to build a module in the circuit and return it.
+/// The name provided is only a namehint for the module---a unique name will be
+/// generated if there are conflicts with the namehint in the circuit-level
+/// namespace.
+FModuleOp LowerGroupsPass::buildNewModule(OpBuilder &builder, Location location,
+                                          Twine namehint,
+                                          SmallVectorImpl<PortInfo> &ports) {
+  llvm::sys::SmartScopedLock<true> instrumentationLock(*circuitMutex);
+  StringRef newModuleName = circuitNamespace->newName(namehint);
```

Comment:
This doesn't seem deterministic.

Reply:
It is not. 😉  How problematic do you think this is?

These module names are, pedantically, not part of the ABI. So, the non-determinism is allowable. However, non-determinism is never great.

The pass can pre-allocate the module names before it enters the parallel region to avoid this. (I'll update this.)

h/t to @uenoku for having the same comment on the original PR: https://github.com/llvm/circt/pull/5536#discussion_r1261067584

Reply:
Updating to make this determinstic in the rebase.